### ISSUE TITLE
Deprecation for private and semi-private callables

### ISSUE THREAD

**Issue by kryzar** [sagemath/sage#38211]
### Problem Description

The Developer Guide, [Section *Deprecation*](https://doc.sagemath.org/html/en/developer/coding_in_python.html#deprecation) does not specify if the deprecation process applies to semi-private (name starts with exactly one underscore) or private (name starts with exactly two underscores) callables (functions, methods, attributes, classes, etc).

### Proposed Solution

I think the following two options are both valid:

- Semi-private or private methods are for devs and are not part of the interface; users should not use them, which is a Python-wide convention, and the devs should not have to worry about keeping compatibility for semi-private and private callables.

- However, it seems the developer guide deprecation process applies to those as well, and not all users are aware of the `_` and `__` convention.

Both options are fine to me and I believe this is a minor issue. However, I think the community should pick a single approach and make it somewhat official but explicitly putting it in the developer guide.

### Alternatives Considered

See above.

### Is there an existing issue for this?

- [X] I have searched the existing issues for a bug report that matches the one I want to file, without success.

**Comment by grhkm21** [comment]
@mkoeppe @tscrim @kwankyu Pinging you three because of the discussion on my PR, also about deprecating things.

Travis has [commented in the past](https://github.com/sagemath/sage/pull/36581#discussion_r1486179644) on my other PR that semi-private and private methods indeed do not go through the deprecation process, but [other developers](https://github.com/sagemath/sage/pull/38199#issuecomment-2164295216) are saying I'm cherry-picking comments, so writing this down somewhere concrete would be better.

**Comment by kwankyu** [comment]
To avoid further confusion, let's try to settle down this issue. Continuing the discussion from #38207, I suggest:

The deprecation policy means what is described in https://doc.sagemath.org/html/en/developer/coding_in_python.html#deprecation.

We want to clarify the policy with the following details:

(S) Only the definitions of functions/methods/attributes/classes living under `sage.` are subject to the deprecation policy.
(M) A function/method/attribute whose name starts with one or two underscores are considered internal, and is not subject to the deprecation policy.
(C) A class whose name has an underscore anywhere is internal, and is not subject to the deprecation policy.

More suggestions?

**Comment by kryzar** [comment]
Sounds good to me. Maybe some examples could be provided in the dev. guide.

**Comment by roed314** [comment]
I worry that (C) is too permissive.  For example, `NumberField_quadratic` has an underscore; I think it's fairly common in the Sage codebase to use underscores to divide sections of a name.

Is there an example of a class that you think should be considered internal?

**Comment by dimpase** [comment]
> I worry that (C) is too permissive. For example, `NumberField_quadratic` has an underscore; I think it's fairly common in the Sage codebase to use underscores to divide sections of a name.

indeed, (C) should not happen.

> 
> Is there an example of a class that you think should be considered internal?

**Comment by kwankyu** [comment]
> For example, `NumberField_quadratic` has an underscore; I think it's fairly common in the Sage codebase to use underscores to divide sections of a name.

Right. Then do you think that `NumberField_quadratic` is in public API? So that if you would change `NumberField_quadratic` to `NumberField_quadratic_extension`,  `NumberField_quadratic` should be subject to the deprecation policy?

If you think so, would you suggest a replacement of (C)?

Or do you think any class name (starting with a letter other than an underscore) change should be subject to the deprecation policy?

**Comment by mkoeppe** [comment]
> For example, `NumberField_quadratic` has an underscore;

Yes, this is a good example for this discussion. `NumberField_quadratic` an implementation class, and it is not necessary to have it in the public API.

The public API to create a number field is through the `sage.rings.number_field.number_field.NumberField` factory function. 
And a way to check whether an object `K` is a quadratic number field using only public API is to use `isinstance(K, sage.rings.number_field.number_field_base.NumberField) and K.base_field == QQ and K.degree() == 2`.

**Comment by roed314** [comment]
Does the deprecation policy apply only to renaming the class, or also to changing methods on it?  I agree that it's probably fine to rename `NumberField_quadratic` to `NumberField_quadratic_extension`; it's not a good idea to be able to modify methods on it at will just because this class is considered "internal."

**Comment by mkoeppe** [comment]
Users should certainly be able to rely on the interface of public methods of objects that have been constructed using public API.

**Comment by saraedum** [comment]
I'm not opposed to the proposal. Just for the record, these classes are in `sage.all` and contain an underscore:

```
['Berkovich_Cp_Affine',
 'Berkovich_Cp_Projective',
 'DynamicalSemigroup_affine',
 'DynamicalSemigroup_projective',
 'DynamicalSystem_Berkovich',
 'DynamicalSystem_affine',
 'DynamicalSystem_product_projective',
 'DynamicalSystem_projective',
 'Euler_Phi',
 'GroupExp_Class',
 'PermutationGroupMorphism_id',
 'PermutationGroupMorphism_im_gens',
 'PermutationGroup_generic',
 'PermutationGroup_subgroup',
 'Riemann_Map']
```

**Comment by kwankyu** [comment]
If we adopt (C), then this gives a clear reason to choose either one of `QuadraticNumberField` (public) and `NumberField_quadratic` (internal) to name a class implementing quadratic number fields. 

Moreover after adopting (C), if one wants to make a public version of `NumberField_quadratic`, then he can just rename it to `QuadraticNumberField` without deprecation warning :-)

**Comment by kwankyu** [comment]
> I'm not opposed to the proposal but just for the record, these classes are in `sage.all` and contain an underscore:
> 
> ```
> ['Berkovich_Cp_Affine',
>  'Berkovich_Cp_Projective',
>  'DynamicalSemigroup_affine',
>  'DynamicalSemigroup_projective',
>  'DynamicalSystem_Berkovich',
>  'DynamicalSystem_affine',
>  'DynamicalSystem_product_projective',
>  'DynamicalSystem_projective',
>  'Euler_Phi',
>  'GroupExp_Class',
>  'PermutationGroupMorphism_id',
>  'PermutationGroupMorphism_im_gens',
>  'PermutationGroup_generic',
>  'PermutationGroup_subgroup',
>  'Riemann_Map']
> ```

We should regard these as exceptions to the rule (C) initially, and sort them our case by case basis with deprecation warnings.

**Comment by roed314** [comment]
How about

(C) A class that is not imported into the global namespace and whose name has an underscore anywhere is internal, and is not subject to the deprecation policy.  This means that the name of the class can be changed, but methods on the class may still be subject to deprecation if instances of the class are returned by Sage's public API.

**Comment by kwankyu** [comment]
> Does the deprecation policy apply only to renaming the class, or also to changing methods on it? 

If (C) applies to a class, the class is internal, and anything with it can be done without deprecation warning, including even removing the class entirely as we refactor code.

As Matthias put it, public methods of the class accessible by public API should still be accessible via other class by public API.

**Comment by mkoeppe** [comment]
Regarding deprecation of things in the public namespace, see
- #25383

**Comment by mkoeppe** [comment]
Another relevant pointer: https://doc.sagemath.org/html/en/developer/packaging_sage_library.html#module-level-runtime-dependencies on the "modularization anti-pattern" `isinstance(..., SomeImplementation_class)`

**Comment by mkoeppe** [comment]
> 'Euler_Phi',

This one was recently garbage-collected

**Comment by tscrim** [comment]
Any class that is in our public *documentation* is subject to be imported and used in a user's code (such as my own that I have/will not make a PR for). So (C) or any variation thereof is a poor option as it is very likely to break code in the wild (without any deprecation message); especially as we are moving to doing `isinstance` checks in our code base (which people use for reference in writing their own).

I think (ME) is the only sensible thing to do. If a user is really using hidden functions, they clearly know what they are doing and understand the risks that it might change without notice. Experimental code explicitly says it can change without notice.

**Comment by mkoeppe** [comment]
> [...] especially as we are moving to doing `isinstance` checks 

We aren't moving to doing `isinstance` checks with random implementation classes. We are moving to doing `isinstance` checks with advertised abstract base classes.

**Comment by tscrim** [comment]
> > [...] especially as we are moving to doing `isinstance` checks
> 
> We aren't moving to doing `isinstance` checks with random implementation classes. We are moving to doing `isinstance` checks with advertised abstract base classes.

Sorry, I was not being precise enough. Yes, we are moving towards that and away from `is_*` functions. The key word is _advertised_, which to me means in the public documentation.

**Comment by mkoeppe** [comment]
Of course these ABCs are in the public documentation.

**Comment by mkoeppe** [comment]
What does not check out is the other direction. Just because an implementation class is documented, it does not mean that it is part of the public API.

**Comment by saraedum** [comment]
> Any class that is in our public documentation is subject to be imported and used in a user's code

I don't think I agree with this argument. A `_method()` is also in our documentation and still we do not tell people that they can rely on its API I think. I would want to use classes as implementation details but if all classes are part of the stable API then we only have `_method()` to hide implementation details?

Is this what you had in mind?

**Comment by tscrim** [comment]
> > Any class that is in our public documentation is subject to be imported and used in a user's code
> 
> I don't think I agree with this argument. A `_method()` is also in our documentation

How are you defining documentation here? In a strict reading, whereby it is a method/function/etc. with a docstring, then yes. Yet, my point is about what is public, and to me, that is what appears in our reference manual as we say that is the official (public) SageMath documentation. That is what you typically find by Google, what I tell people to look at, etc.

> and still we do not tell people that they can rely on its API I think. I would want to use classes as implementation details but if all classes are part of the stable API then we only have `_method()` to hide implementation details?

If you want to hide it, then make it a `_HiddenClass`. If you are going to split our a computational part into a class, then shouldn't users generally have access to that to use it as they see fit?

> Is this what you had in mind?

Does that clarify? Of course, we don't have to add a particular `.py` to the reference manual too, which would make it hidden. Although I think that needs a very high level of justification, but that is a digression.

**Comment by mkoeppe** [comment]
> my point is about what is public, and to me, that is what appears in our reference manual as we say that is the official (public) SageMath documentation. That is what you typically find by Google, what I tell people to look at, etc.

That's much too rigid. Our documentation is written for a wide audience, not just for an audience that is looking for stable public API.

**Comment by kwankyu** [comment]
According to the documentation,

| class name in `sage.all` |  intended to be |
|-------------------------|----------------|
| `Berkovich_Cp_Affine` |  public |
| `Berkovich_Cp_Projective` | public |
| `DynamicalSemigroup_affine` | public |
| `DynamicalSemigroup_projective` | public |
| `DynamicalSystem_Berkovich` | public |
| `DynamicalSystem_affine` | public |
| `DynamicalSystem_product_projective` | internal, accessible via `DynamicalSystem_projective` |
| `DynamicalSystem_projective` | public |
| `Euler_Phi` | internal, accesible via `euler_phi` |
| `GroupExp_Class` | internal, accessible via `GroupExp` |
| `PermutationGroupMorphism_id` | internal, accessible via `PermutationGroupMorphism` |
| `PermutationGroupMorphism_im_gens` | public |
| `PermutationGroup_generic` | internal, accessible via `PermutationGroup` |
| `PermutationGroup_subgroup` | internal, accessible via subgroup method of `PermutationGroup` |
| `Riemann_Map` | public |

**Comment by saraedum** [comment]
Thanks for the clarification.

> If you want to hide it, then make it a _HiddenClass.

In the part of the code base that I've been working on (mostly `sage.rings`) that's not a common practice at all. I would claim that the common practice is what (C) describes. But in other namespaces the practice might be different.

> If you are going to split our a computational part into a class, then shouldn't users generally have access to that to use it as they see fit?

I am not sure I understand what you are trying to say. I am happy for anybody to use my classes as they wish. But I don't want them to think that they are using a stable API.

**Comment by tscrim** [comment]
> > my point is about what is public, and to me, that is what appears in our reference manual as we say that is the official (public) SageMath documentation. That is what you typically find by Google, what I tell people to look at, etc.
> 
> That's much too rigid. Our documentation is written for a wide audience, not just for an audience that is looking for stable public API.

I don't understand what you're trying to say here. What I am saying applies specifically to what people should consider as stable (A)PI; it is irrelevant to this discussion what the entire audience could be.

**Comment by kwankyu** [comment]
Travis, would you formulate your opinion in (CT) as a replacement for (C):

(CT) A class is internal if ... Otherwise it is public.

?

**Comment by tscrim** [comment]
(CT) = (M') A function/method/attribute/class/etc. whose name starts with one or more underscores are considered internal or does not appear in the public documentation (the published html  https://doc.sagemath.org/html/en/ or the corresponding pdf), and is not subject to the deprecation policy.

**Comment by kwankyu** [comment]
Thanks. That is very clear.

**Comment by kwankyu** [comment]
Travis, do you agree with or object to (S)?

I am asking because `sage_docbuild.` appears in the documentation. (S) is intended to exclude it from public API.

**Comment by tscrim** [comment]
Good point, I would then amend my statement to have it be in `sage.*`.

**Comment by kwankyu** [comment]
You don't need to. You just agree with (S).

**Comment by kwankyu** [comment]
> Any class that is in our public _documentation_ is subject to be imported and used in a user's code (such as my own that I have/will not make a PR for). So (C) or any variation thereof is a poor option as it is very likely to break code in the wild (without any deprecation message)...

Aside from "public or internal" controversy,  is there a reliably way to emit a deprecation message for a renamed class? Let's say the code is 
```python
from sage import OldClass
print(isinstance(obj, OldClass))
````
Is there a better way than leaving some comment in `sage.py`:
```
# OldClass is gone. NewClass replaces it.
```
I think the user could just read the file `sage.py`.

@mkoeppe from #38207
> For `isinstance` testing in user code, `lazy_import(..., deprecation=...)` works.

 Does it? Would you elaborate the detail?

**Comment by mkoeppe** [comment]
> @mkoeppe from #38207
> 
> > For `isinstance` testing in user code, `lazy_import(..., deprecation=...)` works.
> 
> Does it? Would you elaborate the detail?

This is a mechanism that was developed for modularization purposes, see https://doc.sagemath.org/html/en/developer/packaging_sage_library.html#module-level-runtime-dependencies

**Comment by kwankyu** [comment]
I am aware of it. But if you use it in `sage.py`, then `from sage import OldClass` in user code won't work, I guess...

**Comment by mkoeppe** [comment]
I don't understand

**Comment by kwankyu** [comment]
In the above toy example, I guess you write something like
```
lazy_import('sage', 'NewClass', 'OldClass',  deprecation=14275)
```
in `sage.py`. Then in user code,`from sage import OldClass` will import `NewClass` as `OldClass`, with a deprecation warning. Is this what you meant?

**Comment by mkoeppe** [comment]
Well, `isinstance(..., OldClass)` works because it delegates to the isinstance test of `NewClass` via `__instancecheck__`

**Comment by kwankyu** [comment]
Ah, okay. It works well!

For real world example, I wrote
```python
from sage.misc.lazy_import import lazy_import
lazy_import('sage.rings.number_field.number_field', 'NumberField_quadratic', 'NumberField_quadratic_extension', deprecation=666)                  
```
into `src/sage/rings/number_field/number_field.py`. Then
```python
sage: from sage.rings.number_field.number_field import NumberField_quadratic_extension  # old class
sage: R = QQ[i]
sage: isinstance(R, NumberField_quadratic_extension)
<ipython-input-3-4851a01f0ac4>:1: DeprecationWarning: 
Importing NumberField_quadratic_extension from here is deprecated; please use "from sage.rings.number_field.number_field import NumberField_quadratic as NumberField_quadratic_extension" instead.
See https://github.com/sagemath/sage/issues/666 for details.
  isinstance(R, NumberField_quadratic_extension)
True
```
though the message is not appropriate in our case.

**Comment by mezzarobba** [comment]
I agree with (SME), and with the fact that (C) is too permissive. I sort of agree with (M'), with the understanding that it only applies to changes that to not affect the usage of objects created using the public API. (E.g., if objects of a private class C can be created using the public API, it is often okay to rename C itself but not its public methods, even if they happen no to appear in the manual.)

In many cases, however, the best policy I can think of is “use common sense”. Some underscore-methods are hard to to without for performance reasons, or are widely used in Sage modules not directly related to the one where they are defined. This is an indication that they are de facto semi-public; be nice to your users and avoid removing them without deprecation if you can. On the other hand, some of the classes targeted by (C) are really unlikely to ever have been used anywhere non-obvious: it may be fine to refactor or delete them without formal deprecation, especially when it would be inconvenient to do otherwise. (Also, some parts of Sage are just so fundamentally broken that is is basically impossible to fix them without breaking some use cases. Do your best then to preserve the ones that actually matter to people!)

**Comment by kwankyu** [comment]
> Aside from "public or internal" controversy, is there a reliably way to emit a deprecation message for a renamed class? 

(RNC) `NumberField_quadratic_extension` is renamed to `NumberField_quadratic`, and `NumberField_quadratic_extension` is to be deprecated. Put
```python
from sage.misc.lazy_import import lazy_import
lazy_import('sage.rings.number_field.number_field', 'NumberField_quadratic', 'NumberField_quadratic_extension', deprecation=666)                  
```
into `src/sage/rings/number_field/number_field.py`. A year later, the `lazy_import` statement is removed.

(RMC) To remove `NumberField_quadratic`  class: (Perhaps a rare case if ever.) We first rename `NumberField_quadratic' to ``NumberField_quadratic_deprecated', and put
```python
from sage.misc.lazy_import import lazy_import
lazy_import('sage.rings.number_field.number_field', 'NumberField_quadratic_deprecated', 'NumberField_quadratic', deprecation=666)                  
```
into `src/sage/rings/number_field/number_field.py`. A year later, the `lazy_import` statement is removed together with  `NumberField_quadratic_deprecated` class.

For each of (RNC) and (RMC), we may have customized "`lazy_import`" command that emits appropriate deprecation message.

Do we already have precedents of (RNC) and (RMC)?

**Comment by tscrim** [comment]
@mezzarobba I would say a class with a leading underscore that can enter through the public API (especially without any special imports) must appear in the public documentation. All a user has to do is `X?` to see the doc, so it really should be searchable online from the `type(X)`. I somewhat agree with the “use common sense” principle, but I would favor deprecations in most cases because we just don’t know what will appear so much in the wild.

**Comment by kwankyu** [comment]
(SME) = (S) and (M) and (E)

**Comment by kwankyu** [comment]
> In many cases, however, the best policy I can think of is “use common sense”. 

I am curious what your "common sense" tells you about the particular case of renaming `PolynomialRing_general` to `Polynomial_generic` (the issue of #38207). Should we deprecate or not?

**Comment by mezzarobba** [comment]
> (SME) = (S) and (M) and (E)

To clarify, I agree that (S), (M), and (E) all hold, i.e., any of (S), (M), and (E) is a valid reason to make breaking changes without deprecation.

**Comment by mezzarobba** [comment]
> @mezzarobba I am curious what your "common sense" tells you about the particular case of renaming `PolynomialRing_general` to `Polynomial_generic` (the issue of #38207). Should we deprecate or not?

Why not just keep it as an alias for the new name (without even deprecating it for now)?

But otherwise, yes, I think we should deprecate it before removing it. I see at least two valid use cases for it (testing if something is a univariate polynomial ring, or creating custom univariate polynomial rings with one's own element class), plus one that cannot reasonably be covered by a deprecation policy but that we should try not to break anyway (bypassing the very slow `PolynomialRing` constructor).

**Comment by kwankyu** [comment]
@saraedum Then (C), which is a simple rule unlike "common sense", would not work for you.

**Comment by kwankyu** [comment]
Of course, the "common sense" is problematic when people's opinions differ. Already for `PolynomialRing_general`, my common sense tells me that it is internal :-)

**Comment by kwankyu** [comment]
Another possible rule:

(CL) A class is internal if it starts with an underline or it cannot be importable from command line

**Comment by saraedum** [comment]
> @saraedum Then (C), which is a simple rule unlike "common sense", would not work for you.

I am happy with (C) but if this is not how some parts of the codebase are written, then I would not insist on it at all.

Here's another example of "common sense". Maybe @roed314 can add more context here, but in some p-adic refactorings  not being able to apply common sense would be really annoying. There are all these extension classes in `padic_extension_leaves.py` that are hidden behind factories. I cannot imagine that anybody would instantiate these classes without the factories since coming up with all the parameters is quite complicated. And also an `isinstance` check with these classes seems quite unlikely because the class names are unwieldy and the things you care about in applications are all exposed in convenient `is_…` methods.

In that example, I would be happy to have some gadget floating around that spits out an error if you try to use the old name, but providing deprecation for these classes and their exact `__init__` interface, seems like a lot of effort for no practical benefit at all?

**Comment by saraedum** [comment]
> Another possible rule close to "common sense":
> 
> (CL) A class is internal if its name starts with an underline or it cannot be importable from command line

Does this mean (C) + `cdef` classes? I think `cdef` is more complicated anyway, not sure we want to start this discussion here as well. With `cdef` there is also the question of ABI compatibility (e.g., 9.2 had a breaking change in the ABI of `element.pxd`.)

**Comment by kwankyu** [comment]
> I am happy with (C) but if this is not how some parts of the codebase are written, then I would not insist on it at all.

Are you thinking of classes that have no underline in it, but are intended to be internal?

**Comment by saraedum** [comment]
> > I am happy with (C) but if this is not how some parts of the codebase are written, then I would not insist on it at all.
> 
> Are you thinking of classes that have no underline in their names, but are intended to be internal?

No sorry, I am personally fine with adopting (C) but I understand that this might be too liberal for some. I understand that @tscrim thinks that (C) is too liberal, see https://github.com/sagemath/sage/issues/38211#issuecomment-2177328302.

And in some cases, it's too strict, as I tried to show in the example in https://github.com/sagemath/sage/issues/38211#issuecomment-2178839080.

I might be misunderstanding something here, but I don't think that just prefixing classes whose "name" is internal with a `_` solves the problem here. Often a class whose name should not be used is still returned to the user, e.g., through a factory. Its methods are therefore part of the public API and I would want their docstrings to show up in our documentation; if we prefix the class name with a `_` that's not happening anymore, right?
Here's a random example, I would consider the name `sage.rings.padics.padic_ZZ_pX_CR_element.pAdicZZpXCRElement` not public, but its method [`expanson`](https://doc.sagemath.org/html/en/reference/padics/sage/rings/padics/padic_ZZ_pX_CR_element.html#sage.rings.padics.padic_ZZ_pX_CR_element.pAdicZZpXCRElement.expansion) should show up in the documentation.

**Comment by mkoeppe** [comment]
+1 on common sense ... and trust in the domain experts ... over any rigid rules.

- Example: In #38228, we decided that `QuaternionAlgebraFactory` (although there is no underscore in sight) can be considered internal and does not need to go through deprecation.

**Comment by kwankyu** [comment]
> I might be misunderstanding something here, but I don't think that just prefixing classes whose "name" is internal with a `_` solves the problem here. ... Here's a random example, I would consider the name `sage.rings.padics.padic_ZZ_pX_CR_element.pAdicZZpXCRElement` not public, ...

By (C), `padic_ZZ_pX_CR_element.pAdicZZpXCRElement` is internal (not public) because there is an underline in the middle of the name. No need to prefix it with an underscore.

By (CT) = (M') , `padic_ZZ_pX_CR_element.pAdicZZpXCRElement` is public. To make this internal, you have to prefix it with an underscore. Then you would have a trouble with the documentation.

By (CL), it is public.

> And in some cases, it's too strict ...

If you mean (C), then Matthias' example `QuaternionAlgebraFactory` would be the case.

**Comment by kwankyu** [comment]
Here is a version of "common sense" rule:

(CS) A class is internal if experts (reviewers for a PR) think so. If experts disagree, then it is public.

**Comment by kwankyu** [comment]
@mkoeppe Would (RNC) and (RMC) in https://github.com/sagemath/sage/issues/38211#issuecomment-2178424246 be solutions to deprecate renamed or removed public (not internal) class?

**Comment by roed314** [comment]
> > I might be misunderstanding something here, but I don't think that just prefixing classes whose "name" is internal with a `_` solves the problem here. ... Here's a random example, I would consider the name `sage.rings.padics.padic_ZZ_pX_CR_element.pAdicZZpXCRElement` not public, ...
> 
> By (C), `padic_ZZ_pX_CR_element.pAdicZZpXCRElement` is internal (not public) because there is an underline in the middle of the name. No need to prefix it with an underscore.

The underscores are in the filename, so they don't count.  I believe that it's public under (C).

**Comment by kwankyu** [comment]
Ah, I see. I missed the dot :-)

**Comment by mkoeppe** [comment]
> Would (RNC) and (RMC) in [#38211 (comment)](https://github.com/sagemath/sage/issues/38211#issuecomment-2178424246) be solutions to deprecate renamed or removed public (not internal) class?

Any solution using `lazy_import` only works for `isinstance`, but not for subclassing, as I explained in https://github.com/sagemath/sage/pull/38207#discussion_r1638770236

**Comment by kwankyu** [comment]
Would subclassing work with imported lazy_import object?

**Comment by mkoeppe** [comment]
No, because a `LazyImport` instance is not even a subclass of `type`.

**Comment by kwankyu** [comment]
So still (RNC) will break the user code... This is better than no deprecation message. ?

**Comment by mkoeppe** [comment]
The primary purpose of deprecation is compatibility, not sending messages.

It will again depend on common sense / domain expertise to determine whether compatibility for subclassing should be sacrificed just for issuing a deprecation message.

**Comment by kwankyu** [comment]
For removed public class, (RMC) would work? But this is a rare (if ever) case...

**Comment by kwankyu** [comment]
> For a removed (rather to be removd) public class, (RMC) would work? But this is a rare (if ever) case...

Again no. Would not work for subclassing...

**Comment by kwankyu** [comment]
> The primary purpose of deprecation is compatibility, not sending messages.

The primary purpose is rather sending message while keeping compatibility. The intention is to notify the user with what would happen before it happens. "Incompatibility and message" seems better than "compatibility and no message". 

and I think (RNC) is a good solution for all other cases.

**Comment by mkoeppe** [comment]
> "Incompatibility and message" seems better than "compatibility and no message".

Absolutely not.

**Comment by kwankyu** [comment]
"compatibility and no message" is just a delayed break (with surprise) of user code. No?

**Comment by mkoeppe** [comment]
That's right. Delaying a breaking change is the point of deprecation. The delay gives people time to act.

**Comment by tscrim** [comment]
The point of the deprecation is to emit a message so users are not surprised when it does go away. Likewise, if it is incompatible (which can be possible) but it does give a message, then that can make clear to the user that they need to change something at that time.

**Comment by mkoeppe** [comment]
> Likewise, if it is incompatible (which can be possible) but it does give a message, then that can make clear to the user that they need to change something at that time.

That's not deprecation, that's an error message.

**Comment by kwankyu** [comment]
> That's right. Delaying a breaking change is the point of deprecation. The delay gives people time to act.

If people is not notified somehow, what is the use of the delay?

**Comment by mkoeppe** [comment]
Deprecation messages are just one part of a communication strategy. Downstream developers stay informed about the projects that they use in various ways.

**Comment by tscrim** [comment]
> > Likewise, if it is incompatible (which can be possible) but it does give a message, then that can make clear to the user that they need to change something at that time.
> 
> That's not deprecation, that's an error message.

Not in general. Plus, you need to be able to set the message, which you cannot do with, e.g., an `ImportError`.

**Comment by kwankyu** [comment]
It is getting vague... 

If "compatibility and (in-code) message" is not possible (as in the subclassing case), then some message (communication) should be posted somewhere at least. We may post these in the release note. I think these deprecations of public classes are rare events. So this way of communication would not be burdensome...

**Comment by mkoeppe** [comment]
> If "compatibility and (in-code) message" is not possible (as in the subclassing case), then some message (communication) should be posted somewhere at least. We may post these in the release note.

Well, deprecations should be noted already in the title of the PR that deprecates it, which will make it show up in the automatically generated release notes: https://github.com/sagemath/sage/releases/tag/10.1

Having a section on deprecations in release notes is a very standard practice. See for example https://docs.scipy.org/doc/scipy/release/1.13.0-notes.html#deprecated-features

**Comment by mkoeppe** [comment]
Also https://github.com/sagemath/sage/wiki/Sage-10.4-Release-Tour#deprecations

**Comment by kwankyu** [comment]
OK. It seems that posting deprecations of public classes in the "Deprecation" section of the release notes is the best we can do. So deprecation of public classes is done differently from the deprecation policy of functions/methods/attributes.

**Comment by mkoeppe** [comment]
Note that also for some aspects of our Cython interfaces we do not have any deprecation infrastructure other than code comments and release notes.

**Comment by kwankyu** [comment]
Back to "public and internal" discussion,

we now have, as alternatives:

(C) A class is internal if its name has an underscore anywhere is internal

(CT) = (M') A function/method/attribute/class/etc. whose name starts with one or more underscores are considered internal or does not appear in the public documentation (the published html https://doc.sagemath.org/html/en/ or the corresponding pdf), and is not subject to the deprecation policy.

(CL) A class is internal if its name starts with an underline or it cannot be importable from command line

(CS) A class is internal if experts (reviewers for a PR) think so. If experts disagree, then it is public.

People here give their preferences or suggest modifications or a new alternative.

I will collect a couple of most preferred alternatives, and bring the discussion to sage-devel.

**Comment by saraedum** [comment]
Thanks for compiling the list. I like (CL) + (CS).

**Comment by roed314** [comment]
I agree with (CL) + (CS).  We could elaborate slightly on (CS) by saying
(CS') A class is internal if experts (reviewers for a PR) agree that it is unlikely to be directly imported by user code. If experts disagree, then it is public.

**Comment by grhkm21** [comment]
Is my understanding correct that (CT) = (M') is a subset of (CL)?

Anyways, I am okay with (CT) + (CL) + (CS)/(CS'). FWIW (CS) has been the case [in the past](https://github.com/sagemath/sage/pull/37346#issuecomment-1945380519)

**Comment by kwankyu** [comment]
> Is my understanding correct that (CT) = (M') is a subset of (CL)?

Actually as sets, probably they are the same, unless there is some module that is omitted in the documentation (perhaps by mistake).

**Comment by kwankyu** [comment]
We may define the following standard procedures, from practices.

(RNP) *Procedure to rename a public class*
(RNP step 1) rename `class OldClass` to `class NewClass`, add alias `OldClass = NewClass`:
```python
class NewClass:
    ....

OldClass = NewClass   # OldClass is deprecated. See Issue 12345.
```
(RNP step 2) before the next stable release, post the change to the "Deprecated" section of the release note.

The alias will be there forever unless someone steps up to remove it :-) 

(RMP) *Procedure to remove a public class*
(RMP step 1) add a comment
```python
# Class is deprecated. See Issue 12345.

class Class:  
```
(RMP step 2) before the next stable release, post the change to the "Deprecated" section of the release note.

**Comment by kwankyu** [comment]
I think we are already converging to a consensus. So I will bring only the following items to sage-devel for wider discussion:

(S) Only public functions, methods, attributes, classes living under `sage` are subject to the deprecation policy.

(M) Functions, methods, attributes, classes whose names start with an underscore are considered internal, and is not subject to the deprecation policy.

(C) A class whose name does not start with an underscore is public by default, but can be considered internal if experts agree that the class is unlikely to be directly imported by user code. If experts disagree, then it is public. For a PR changing the class, authors and reviewers are the experts.

**Comment by vincentmacri** [comment]
I think this was closed by #38273.
